predicators/approaches/grammar_search_invention_approach.py

```diff
@@ -513,56 +572,135 @@ class _FeatureDiffInequalitiesPredicateGrammar(
         _SingleFeatureInequalitiesPredicateGrammar):
     """Generates features of the form "|0.feature - 1.feature| <= c"."""
 
+    def _yield_pred_given_const(
+            self, feature_ranges: Dict[Type, Dict[str, Tuple[float, float]]],
+            constant_idx: int, constant: float,
+            cost: float) -> Iterator[Tuple[Predicate, float]]:
+        for (t1, t2) in itertools.combinations_with_replacement(
+                sorted(self.types), 2):
+            for f1 in t1.feature_names:
+                for f2 in t2.feature_names:
+                    # To create our classifier, we need to leverage the
+                    # upper and lower bounds of its features.
+                    # First, we extract these and move on if these
+                    # bounds are relatively close together.
+                    lb1, ub1 = feature_ranges[t1][f1]
+                    if abs(lb1 - ub1) < 1e-6:
+                        continue
+                    lb2, ub2 = feature_ranges[t2][f2]
+                    if abs(lb2 - ub2) < 1e-6:
+                        continue
+                    lb, ub = utils.compute_abs_bounds_given_frange(
+                        lb1, ub1, lb2, ub2)
+                    # Scale the constant by the correct range.
+                    k = constant * (ub - lb) + lb
+                    # Create classifier.
+                    comp, comp_str = le, "<="
+                    diff_classifier = _AttributeDiffCompareClassifier(
+                        0, t1, f1, 1, t2, f2, k, constant_idx, comp, comp_str)
+                    name = str(diff_classifier)
+                    types = [t1, t2]
+                    pred = Predicate(name, types, diff_classifier)
+                    assert pred.arity == 2
+                    yield (pred, cost)
+
     def enumerate(self) -> Iterator[Tuple[Predicate, float]]:
         # Get ranges of feature values from data.
         feature_ranges = self._get_feature_ranges()
         # Edge case: if there are no features at all, return immediately.
         if not any(r for r in feature_ranges.values()):
             return
+        # Start by generating predicates such that the two features are
+        # very close together.
+        for ret_val in self._yield_pred_given_const(feature_ranges, 0,
+                                                    (1 / 60.0), 4.0):
```

Comment:
where are these numbers (1/60 and 4) coming from?

Reply:
maybe have the small constant `epsilon=1/60.0` come from `settings.py`?

Reply:
Pulling them into settings is good, but I also want to understand for myself why the magic is needed. If the intention is to choose a value that measures "closeness", why wouldn't something like 1e-6 work?

Reply:
oh - good question! It's because objects have some amount of "size", and so even when they're touching, it's not like their centers overlap. E.g. in stick button, when the robot touches the button, the center of the robot and the object might still be offset by a bit

Reply:
Got it, thanks! Let's leave a note somewhere for future users. That would be easy to get confused about when trying this grammar in a new environment.